dbms/src/Interpreters/ExpressionActions.cpp

```diff
@@ -1076,10 +1076,25 @@ void ExpressionActionsChain::finalize()
     for (int i = static_cast<int>(steps.size()) - 1; i >= 0; --i)
     {
         Names required_output = steps[i].required_output;
+        std::unordered_map<String, size_t> required_output_indexes;
+        for (size_t j = 0; j < required_output.size(); ++j)
+            required_output_indexes[required_output[j]] = j;
+        auto & can_remove_required_output = steps[i].not_need_in_future_steps;
+
         if (i + 1 < static_cast<int>(steps.size()))
         {
+            const NameSet & additional_input = steps[i + 1].additional_input;
```

Comment:
Is it ok that we are considering only the next step when setting the `not_need_in_future_steps` variable (the name points to all future steps instead of just the next one).

Also it's better to stick to a single name (`can_remove_required_output` or `not_need_in_future_steps`)